File: add-two-numbers/ulises/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *r = new ListNode();
        ListNode *p;
        p = r;
        int l = 0;        
        while(true)
        {
            int v1 = l1 == nullptr ? 0 : l1->val;
            int v2 = l2 == nullptr ? 0 : l2->val;
            int s = l + v1 + v2;
            p->val = s % 10;
            l = s > 9 ? 1 : 0;
            l1 = l1 == nullptr ? nullptr : l1->next;
            l2 = l2 == nullptr ? nullptr : l2->next;
            if(l1 != nullptr || l2 != nullptr || l != 0)
            {
                p->next = new ListNode();
                p = p->next;
            }
            else
            {
                break;
            }
        }
        return r;
    }
};

File: linked-list/add-two-numbers/enrique/add-two-numbers.go
package enrique

// ListNode is Definition for a singly-linked list.
type ListNode struct {
    Val int
    Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	solution := &ListNode{}
	carry := false

	sum := solution
	currentL1 := l1
	currentL2 := l2
	for {
		if currentL1 == nil && currentL2 == nil && !carry{
			break
		}

		a, b, c := 0, 0, 0
		if currentL1 != nil {
			a = currentL1.Val
			currentL1 = currentL1.Next
		}

		if currentL2 != nil {
			b = currentL2.Val
			currentL2 = currentL2.Next
		}

		if carry {
			c = 1
		}

		total := a + b + c
		carry = total > 9
		sum.Val = total % 10

		if currentL2 != nil || currentL1 != nil || carry {
			sum.Next = &ListNode{}
			sum = sum.Next
		}
	}

	return solution
}
